std/container/array.d

```diff
@@ -59,87 +48,255 @@ if (!is(Unqual!T == bool))
+                        auto pl = enforce(cast(T*) malloc(sz))[0 .. value];
+                        // copy old data over to new Array
+                        memcpy(pl.ptr, _payload.ptr, T.sizeof * len);
+                        
+                        // Zero out unused capacity to prevent gc from seeing false pointers
+                        if (isGrowing)
+                            initializeAll(pl.ptr[len .. value]);
+                        
+                        GC.addRange(pl.ptr, sz);
+                        GC.removeRange(_payload.ptr);
+                        free(_payload.ptr);
+                        _payload = pl.ptr[0 .. len];
+                    }
+                    else
+                    {
+                        /* These can't have pointers, so no need to zero
+                         * unused region
+                         */
+                        auto pl = enforce(cast(T*) realloc(_payload.ptr, sz))[0 .. value];
+                        if (isGrowing)
+                            initializeAll(pl.ptr[len .. value]);
+                        _payload = pl.ptr[0 .. len];
+                    }
+                }
+            }
+            
+            _capacity = value;
+        }
+        
+        // Requests the Array to change its capacity to at least howMuch.
+        // If howMuch is greater than the current capacity the Array is
+        // resized, otherwise no action is taken and the capactity
+        // remains unchanged.
+        void reserve(size_t howMuch)
+        {
+            immutable cap = capacity;
+            
+            if (howMuch > cap)
+            {
+                size_t sz = (cap == 0) ? 4 : (cap * 3 / 2); // grow rule 1/3
+                if (sz < howMuch)
+                    sz = howMuch;
+                capacity = sz;
+                
+                assert(capacity > length && _payload.ptr);
+            }
+        }
+        
+        // Requests the Array to reduce its capacity to howMuch.
+        // If the Array length is greater than howMuch the length
+        // will be truncated, before the capacity is reduced.
+        void shrink(size_t howMuch)
         {
-            if (elements <= capacity) return;
-            immutable sz = elements * T.sizeof;
-            static if (hasIndirections!T)       // should use hasPointers instead
+            immutable cap = capacity;
+            
+            if (howMuch < cap)
+            {
+                length = cap;
+                capacity = cap;
+                
+                assert(capacity == length && _payload.ptr);
+            }
+        }
+        
+        // Remove one entry at the specified index.
+        // An entry will be destroyed it appropriately.
+        
+        // Complexity: $(BIGOH n)
+        void removeAt(size_t index)
+        {
+            enforce(index >= 0 && index <= length);
+            static if (is(T == struct) && hasElaborateDestructor!T)
```

Comment:
Don't check for struct. Other types, such as static arrays, can also have destructors. Also, enums of structs I think don't match for struct unless you do `BaseType!T`. Long story short, just trust `hasElaborateDestructor` to know what it's doing. This is a compile time static if, so it's "free".